content/en/blog/_posts/protect-mission-critical-pods-priorityclass/index.md

```diff
@@ -0,0 +1,283 @@
+
+## Resource Management in Kubernetes 
+
+The controlplane consists of multiple components, out of which the scheduler (aka kube-scheduler) is one of the components which is responsible for assigning a node to a pod.
+- Whenever a pod is created, it goes to a pending state, and then the scheduler kicks in and determines which node should be the best-suited node for the placement of the new pod.
+  - In the background, the scheduler runs an infinite loop that looks out for all the pods where the nodeName field is null. When it finds one, it starts the process of finding which node should be the best one. Though there are various parameters for it, just to name a few, they are “nodeName”, “nodeSelector”, “taint and tolerations”, "nodeAffinity", and the rank of available nodes based on available resources in terms of CPU and memory, etc which decides the pod’s placement.
+- If in-case the scheduler is not able to find any node, the pod remains to be in pending state, which is not an ideal scenario.
+
+The below diagram explains how the request flows in (starting from point number 1 to 4):
+
+![Whenever there is a new incoming pod, this is how kube-scheduler schedules a pod.](kube-scheduler.svg)
+
+## Typical Use Cases
+
+1. Let's say the pod which you are planning to deploy is really critical, and you need to run it anyhow. How will you make the resources available in the cluster? By first identifying which pods are not required, then manually deleting them, etc., which is a manual and tedious process, or you may plan to add a new node to the cluster, which too might take some time and money as well.
+An example of such a case is a DaemonSet of an [infrastructure component like Grafana Loki](https://www.infracloud.io/blogs/grafana-loki-log-monitoring-alerting/), the Loki pods must run on each node before any other pods run on that node.
+
+2. Another use case can be: Suppose, you have a single cluster, and that cluster holds the pods for the below environments:
+   - preprod [Priority 1] 
+   - staging [Priority 2] 
+   - development [Priority 3]
+
+   Unfortunately, the cluster goes through high resource consumption, and now there is a resource crunch on the nodes with respect to the CPU and the memory, and we don’t have further nodes to expand or scale our Kubernetes cluster. Either the pods will remain in a pending state or the cluster will start behaving abnormally and if at all eviction starts, it will do that randomly because it doesn’t know which one to evict and which one to keep. 
```

Comment:
```suggestion
   Unfortunately, the cluster goes through high resource consumption, and now there is a resource competition on the nodes with respect to the CPU and the memory. At least in the short term, you don’t have further nodes to expand or scale that Kubernetes cluster (cluster level autoscaling _may_ add more nodes, but this takes time). Perhaps some pods will remain in a Pending state, or perhaps service will degrade as Pods compete for resources. If the kubelet does evict a Pod from a node, that eviction happens at random because the kubelet doesn’t have any special information about which Pods to evict and which to keep.
```